Added email confirmation functionality. Needs designer to refine added pages.

Greengrer opened the pull request:
Now when user is registered they recieve email with link.
![зображення](https://user-images.githubusercontent.com/51015697/120508092-18b41100-c3d0-11eb-9468-5d1c63e3bea2.png)

Link is: app + \emailConfirmation?token= + their confirmation token from db.
![зображення](https://user-images.githubusercontent.com/51015697/120508185-2ec1d180-c3d0-11eb-98a0-03f8b2db684e.png)

When user follows the link they see login page. They have to give correct email and password to become verified and get session.
![зображення](https://user-images.githubusercontent.com/51015697/120508452-66c91480-c3d0-11eb-91de-b23fb32aaeb6.png)

If email isn't correct, loads page confirmation-directions with message asking to log in again.
![зображення](https://user-images.githubusercontent.com/51015697/120508757-a8f25600-c3d0-11eb-8cca-d09202b539ee.png)

If email is correct, but passwor isn't, user stays on confirmation login page with message asking to try again.
![зображення](https://user-images.githubusercontent.com/51015697/120508595-87916a00-c3d0-11eb-9c23-d6e893fd48db.png)

If there token from link can't be converted to UUID or doesn't exist in db, confirmation-directions page loads with message "Sorry, such confirmation link does not exist."
![зображення](https://user-images.githubusercontent.com/51015697/120508966-d3dcaa00-c3d0-11eb-949b-c219ed12620a.png)



1. Added ConfirmationToken.
2. Added database migration file V__1_7.
3. Added ConfirmationTokenDao and tests.
4. Changed DatabasePreparerForITests, DataGeneratorForTests, DataFinderFromDB to work with confirmation tokens and update user status in db.
5. Changed entity User to have email always in lowercase there.
6. Changed classes to work with email case-insensitively.
7. Added ConfirmationService and tests.
8. Fixed bug with sending email: changed port to 587, added "...starttls.enable" - "true", changed auth argument in Starter from propertiesReader.getProperty("mail.smtp.username") to propertiesReader.getProperty("mail.smtp.auth").
9. Changed SignUpServlet to show "confirmation-directions" page and send email with link after successful registration.
10. Changed SignInServlet to show "confirmation-directions" page and send email with link after logging in with correct credantials, but email isn't verified.
11. Changed SignUpServletTest.
12. Changed SignInServletTest.
13. Changed UserDao and tests to have findById().
14. Changed UserService and tests to have findById();
15. Created "confirmation-signin.html" and "confirmation-directions.html" templates.
16. Minor correction in signin.html.
17. Created ConfirmationServlet and tests.
18. Mapped ConfirmationServlet to /emailConfirmation in Starter.
19. Minor refactoring across the project, like removal of extra empty row.

Greengrer committed:
Added email_confirmation functionality.
1. Added ConfirmationToken.
2. Added database migration file V__1_7.
3. Added ConfirmationTokenDao and tests.
4. Changed DatabasePreparerForITests, DataGeneratorForTests, DataFinderFromDB to work with confirmation tokens and update user status in db.
5. Changed entity User to have email always in lowercase there.
6. Changed classes to work with email case-insensitively.
7. Added ConfirmationService and tests.
8. Fixed bug with sending email: changed port to 587, added "...starttls.enable" - "true", changed auth argument in Starter from propertiesReader.getProperty("mail.smtp.username") to propertiesReader.getProperty("mail.smtp.auth").
9. Changed SignUpServlet to show "confirmation-directions" page and send email with link after successful registration.
10. Changed SignInServlet to show "confirmation-directions" page and send email with link after logging in with correct credantials, but email isn't verified.
11. Changed SignUpServletTest.
12. Changed SignInServletTest.
13. Changed UserDao and tests to have findById().
14. Changed UserService and tests to have findById();
15. Created "confirmation-signin.html" and "confirmation-directions.html" templates.
16. Minor correction in signin.html.
17. Created ConfirmationServlet and tests.
18. Mapped ConfirmationServlet to /emailConfirmation in Starter.
19. Minor refactoring across the project, like removal of extra empty row.

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/service/impl/DefaultCaptchaService.java:
```diff
@@ -20,7 +20,7 @@
             return jpegOutputStream.toByteArray();
         } catch (Exception e) {
-            log.error("Error during captcha generation!", e);
+            //log.error("Error during captcha generation!", e);
```
No commented code at develop branch pls

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/web/servlets/ImageCaptchaServlet.java:
```diff
@@ -7,13 +7,14 @@
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
         String captchaId = request.getSession().getId();
-
+        System.out.println("captcha get. " + Instant.now());
```
Please use logger instead sout

Greengrer committed:
Changes after review.
1. Fixed a bug in doPost.
2. Hopefully, fixed a bug with Travis tests erroring because thymeleaf can't load resources/templates/confirmation-signin.html.
3. Uncommented log and deleted cout.

Greengrer committed:
Trying to fix thymeleaf bug for travis.
1. Renamed confirmation-signIn.html to lowercase version: -signin.html.
2. Made file names in ConfirmationServlet lowercase again.

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/web/servlets/ConfirmationServlet.java:
```diff
@@ -0,0 +1,111 @@
+                    userService.update(user.getEmail(), user.getEmail(), password, user.getDiscogsUserName());
+                    confirmationService.deleteByUserId(user.getId());
+                    HttpSession session = request.getSession(true);
+                    session.setMaxInactiveInterval(60 * 60 * 5);
```
It is looks like magic number. Please add it to application.properties and to this class fields.

Greengrer replied:
*it looks.
Okay!

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcConfirmationTokenDao.java:
```diff
@@ -0,0 +1,122 @@
+        }
+        log.debug("Resulting optional with confirmation token is {'Optional.ofNullable(confirmationToken)':{}}", Optional.ofNullable(confirmationToken));
+        return Optional.ofNullable(confirmationToken);
+    }
```
I think last two rows inside block try could be more readable, and ConfirmationToken confirmationToken = null; also can be placed inside try

Greengrer replied:
Which two rows are you referring to?
We can't place confirmationToken = null inside try: if try block errors, confirmationToken wouldn't be initialised, because everything that is initialised within try {} would be gone.

Yuliia-Sh replied:
47, 48 I would put inside try

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcConfirmationTokenDao.java:
```diff
@@ -0,0 +1,122 @@
+            if (resultInsert > 0) {
+                isAdded = true;
+            }
+        } catch (SQLException e) {
```
I would say here return (resultInsert > 0) and log in the next level if it is needed, but it is just suggestion, I think it can be more readable and understandable

Greengrer replied:
Might be!
But I think that it is dangerous to place logs to next levels: you'd need both dao method and service method to work correctly to receive correct log about token added to database. And the if with logs would stay the same, just in the service method.

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDao.java:
```diff
@@ -16,17 +17,20 @@
             " SET email = ?, password = ?, salt = ?, iterations = ?, role = ?, status = ?, discogs_user_name = ?" +
-            " WHERE email = ?";
+            " WHERE email ILIKE ?";
 
```
I think all quiries should be either with "public." or all without that, I think without would be better

Greengrer replied:
Roger that!

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/mapper/ConfirmationTokenRowMapper.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public ConfirmationToken mapRow(ResultSet resultSet) {
+        ConfirmationToken confirmationToken = new ConfirmationToken();
+        try {
```
why this row outside try? return also can be placed inside

Greengrer replied:
Thank you! You're right.

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/entity/ConfirmationToken.java:
```diff
@@ -0,0 +1,69 @@
+    private long userId;
+    private UUID token;
+    private Timestamp timestamp;
+
```
Shouldn't here LocalDataTime be used?

Greengrer replied:
Database uses Timestamp - when you use LocalDateTime, you still work with Timestamp from result set and have to convert it in LocalDateTime.
I decided not to bother with it, because there is a literal java.sql.Timestamp class for this.

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/service/impl/DefaultConfirmationService.java:
```diff
@@ -0,0 +1,126 @@
+    private static final String SUBJECT = "Confirm your email on wax-deals.com";
+    private static final String MAIL_BODY_BEGINNING_WITH_LINK = "To confirm click on the link below and log into your account. " +
+            "\n https://wax-deals.herokuapp.com/emailConfirmation?token=";
+    private static final String MAIL_BODY_ENDING = "\n Thank you!\n\nWith kindest regards,\nWax-Deals team";
```
https://wax-deals.herokuapp.com - I think it shouldn't be used here, could you please add it into property file

Greengrer replied:
You mean, load as property?
Like, 



<!--StartFragment-->private static final String MAIL_BODY_BEGINNING_WITH_LINK = "To confirm click on the link below and log into your account. ";<!--EndFragment-->
...
StringBuilder emailBuilder = new StringBuilder();
emailBuilder.append(MAIL_BODY_BEGINNING_WITH_LINK);
if (System.getEnv("env") != null) {
    emailBuilder.append("https://wax-deals.herokuapp.com/emailConfirmation?token=");
} else {
    emailBuilder.append("http://localhost:");
    emailBuilder.append(Starter.PROPERTIES_READER.getProperty("appPort"));
    emailBuilder.append("/emailConfirmation?token=");
}
...?

Yuliia-Sh replied:
link to application should be in the property file

Yuliia-Sh replied:
not port

Greengrer replied:
Got you! I think, it should be decided on a team call.

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/service/impl/DefaultConfirmationService.java:
```diff
@@ -0,0 +1,126 @@
+            IllegalArgumentException e = new IllegalArgumentException();
+            log.error("Id is 0 or less {'userId':{}}", userId, e);
+            throw new RuntimeException(e);
+        }
```
I would write - throw new IllegalArgumentException()
why two rows are needed instead one and one more object?

Yuliia-Sh replied:
better first to check parameters, and then do basic logic:
if (userId<=0) throw exception...
Optional<ConfirmationToken> optionalToken = confirmationTokenDao.findByUserId(userId);
....
as for me it is more correct and understandable

Greengrer replied:
So to move exception to the top, and not wrap remaining logic in else {}?

Yuliia-Sh replied:
I think you understood idea, just try to beautify code

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/util/MailSender.java:
```diff
@@ -53,7 +59,7 @@ public boolean sendMail(String recipient, String subject, String mailBody) {
-        } catch (MessagingException e) {
+        } catch (Exception e) {
             log.error("Can't send email to recipient : {}  due to error : {}", recipient, e.toString());
             return false;
```
here exception wasn't logged, was it?

Greengrer replied:
It was, those are not my logs.

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/ConfirmationServlet.java:
```diff
@@ -0,0 +1,111 @@
+            log.warn("Invalid email confirmation link token parameter {'token':{}}", tokenAsString, e);
+        }
+        Optional<ConfirmationToken> optionalConfirmationToken = confirmationService.findByToken(token);
+        response.setContentType("text/html;charset=utf-8");
```
what will be in case of exception? 
confirmationService.findByToken(null) -?
is it correct?

Greengrer replied:
If I can't get token from link, I pass null to service, that returns false, that then causes loading "Sorry, such confirmation link doesn't exist").

Yuliia-Sh replied:
it is just not clear here

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/ConfirmationServlet.java:
```diff
@@ -0,0 +1,111 @@
+            if (userByLinkToken.getEmail().equalsIgnoreCase(email)) {
+                String password = request.getParameter("password");
+                Optional<User> optionalUser = userService.signInCheck(email, password);
+                log.debug("Received a optional with User with password verification by the passed " +
```
one servlet method should contain one call to service method

Greengrer replied:
Even when servlet works with two types of data?
Because I can't take confirmationToken or user out from servlet and have it working properly.
And for different entities should exist different services, right?

Yuliia-Sh replied:
One service can use another, in your case userService can use confirmationService - it is fine

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java:
```diff
@@ -52,28 +55,32 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
                     attributes.remove("email");
                     attributes.remove("discogsUserName");
+                    confirmationService.sendMessageWithLinkToUserEmail(userService.findByEmail(email).get());
+                    PageGenerator.getInstance().process("confirmation-directions", attributes, response.getWriter());
```
one servlet - one service call, I think sending email should be added not to servlet, but into service - userService.add(email, password, discogsUserName), perhaps it can be done even easier (without findByEmail)

Greengrer replied:
Are you sure?
How can you have one servlet and one service call when you need to find data for doGet AND write data for doPost?

Yuliia-Sh replied:
sorry, I mean one method

Greengrer replied:
I still need to find data, like email or token, because I need to put it in templates as values. So in doPost I call a method to find data to check if this is correct link, then if it is, find user by that token, then call method to check if login is successful, then call method to update user, then call method to delete token.
How can I break this doPost up?

Yuliia-Sh replied:
this is SignUpServlet - registration user, am I right?
put confirmationService.sendMessageWithLinkToUserEmail(userService.findByEmail(email).get()); inside userService.add(email, password, discogsUserName)
link always correct here as it is just generated

adding user and adding link can be done in one transaction (if user added, then link also added), when you add user it is possible to return added id from jdbc layer

regarding template you just added (as I can see)
attributes.put("message", "Please confirm your registration. To do this, follow the link that we sent on your email");

Greengrer committed:
Changes after reviews, added property, confirmation token update and tests.

Yuliia08 committed:
tests are fixed

Yuliia08 committed:
refactoring - methods are in the services instead of servlets
